File: src/bentley.ts
import path from "node:path";
import { ContainerClient } from "@azure/storage-blob";
import { getBentleyConfig } from "./config.js";

const ACCEPT_V1 = "application/vnd.bentley.itwin-platform.v1+json";
const ACCEPT_V2 = "application/vnd.bentley.itwin-platform.v2+json";

export interface RealityData {
  id: string;
  displayName?: string;
  authoring?: boolean;
  [key: string]: unknown;
}

export type RealityModelingJobState =
  | "Queued"
  | "Active"
  | "TerminatingOnCancel"
  | "TerminatingOnFailure"
  | "Cancelled"
  | "Failed"
  | "Success";

export interface RealityModelingJob {
  id: string;
  state: RealityModelingJobState;
  type: string;
  iTwinId: string;
  name?: string;
  specifications: Record<string, unknown>;
  executionInfo?: Record<string, unknown>;
  [key: string]: unknown;
}

export interface RealityModelingJobProgress {
  percentage?: number;
  state?: RealityModelingJobState;
}

export class BentleyClient {
  private readonly apiBaseUrl: string;

  public constructor(private readonly accessToken: string, apiBaseUrl?: string) {
    if (!accessToken.trim()) throw new Error("Bentley access token is empty.");
    this.apiBaseUrl = apiBaseUrl?.replace(/\/$/, "") || getBentleyConfig().apiBaseUrl;
  }

  public async createImageCollection(itwinId: string, displayName: string): Promise<RealityData> {
    const payload = await this.request("/reality-management/reality-data/", {
      method: "POST",
      body: JSON.stringify({
        iTwinId: itwinId,
        displayName,
        classification: "Undefined",
        type: "CCImageCollection",
        authoring: true,
      }),
    });

    return realityDataFrom(payload, "Create reality data response did not include an id.");
  }

  public async getWriteContainerUrl(realityDataId: string, itwinId: string): Promise<string> {
    const params = new URLSearchParams({ iTwinId: itwinId });
    const payload = await this.request(
      `/reality-management/reality-data/${encodeURIComponent(realityDataId)}/writeaccess?${params}`,
    );

    const container = objectValue(payload.container) ?? payload;
    const links = objectValue(container._links);
    const containerUrl = objectValue(links?.containerUrl);
    const href = containerUrl?.href;
    if (typeof href !== "string" || !href) {
      throw new Error("Write-access response did not include a container URL.");
    }
    return href;
  }

  public async finalizeRealityData(realityDataId: string, itwinId: string): Promise<RealityData> {
    const payload = await this.request(
      `/reality-management/reality-data/${encodeURIComponent(realityDataId)}`,
      {
        method: "PATCH",
        body: JSON.stringify({ iTwinId: itwinId, authoring: false }),
      },
    );

    return realityDataFrom(payload, "Finalize response was not a reality data object.");
  }

  public async createRealityModelingJob(
    job: Record<string, unknown>,
  ): Promise<RealityModelingJob> {
    const payload = await this.request(
      "/reality-modeling/jobs",
      { method: "POST", body: JSON.stringify(job) },
      ACCEPT_V2,
    );
    return jobFrom(payload);
  }

  public async getRealityModelingJob(jobId: string): Promise<RealityModelingJob> {
    const payload = await this.request(
      `/reality-modeling/jobs/${encodeURIComponent(jobId)}`,
      {},
      ACCEPT_V2,
    );
    return jobFrom(payload);
  }

  public async getRealityModelingJobProgress(
    jobId: string,
  ): Promise<RealityModelingJobProgress> {
    const payload = await this.request(
      `/reality-modeling/jobs/${encodeURIComponent(jobId)}/progress`,
      {},
      ACCEPT_V2,
    );
    const progress = objectValue(payload.progress);
    if (!progress) throw new Error("Reality Modeling progress response was malformed.");
    return progress as RealityModelingJobProgress;
  }

  private async request(
    endpoint: string,
    init: RequestInit = {},
    accept = ACCEPT_V1,
  ): Promise<Record<string, unknown>> {
    const headers = new Headers(init.headers);
    headers.set("Authorization", `Bearer ${this.accessToken}`);
    headers.set("Accept", accept);
    if (init.body !== undefined) headers.set("Content-Type", "application/json");

    const response = await fetch(`${this.apiBaseUrl}${endpoint}`, {
      ...init,
      headers,
    });

    if (!response.ok) {
      const detail = (await response.text()).trim();
      throw new Error(
        `Bentley API returned ${response.status}: ${detail || response.statusText || "request failed"}`,
      );
    }

    const payload: unknown = await response.json();
    const object = objectValue(payload);
    if (!object) throw new Error("Bentley API returned an unexpected response.");
    return object;
  }
}

export async function uploadFiles(options: {
  containerUrl: string;
  files: string[];
  root: string;
  prefix?: string;
  onUploaded?: (file: string, blobName: string) => void;
}): Promise<number> {
  const container = new ContainerClient(options.containerUrl);
  const resolvedRoot = path.resolve(options.root);
  const prefix = options.prefix?.replace(/^\/+|\/+$/g, "") ?? "";
  let uploaded = 0;

  for (const file of options.files) {
    const resolvedFile = path.resolve(file);
    const relative =
      options.files.length === 1 && resolvedRoot === resolvedFile
        ? path.basename(resolvedFile)
        : path.relative(resolvedRoot, resolvedFile);

    if (!relative || relative.startsWith("..")) {
      throw new Error(`Photo is outside upload root: ${resolvedFile}`);
    }

    const normalized = relative.split(path.sep).join("/");
    const blobName = prefix ? `${prefix}/${normalized}` : normalized;
    const blob = container.getBlockBlobClient(blobName);
    await blob.uploadFile(resolvedFile, { conditions: { ifNoneMatch: "*" } });
    uploaded += 1;
    options.onUploaded?.(resolvedFile, blobName);
  }

  return uploaded;
}

function realityDataFrom(payload: Record<string, unknown>, error: string): RealityData {
  const realityData = objectValue(payload.realityData) ?? payload;
  if (typeof realityData.id !== "string" || !realityData.id) throw new Error(error);
  return realityData as RealityData;
}

function jobFrom(payload: Record<string, unknown>): RealityModelingJob {
  const job = objectValue(payload.job) ?? payload;
  if (
    typeof job.id !== "string" ||
    typeof job.state !== "string" ||
    typeof job.type !== "string" ||
    typeof job.iTwinId !== "string" ||
    !objectValue(job.specifications)
  ) {
    throw new Error("Reality Modeling response did not include a valid job.");
  }
  return job as unknown as RealityModelingJob;
}

export function objectValue(value: unknown): Record<string, unknown> | undefined {
  return value !== null && typeof value === "object" && !Array.isArray(value)
    ? (value as Record<string, unknown>)
    : undefined;
}

File: src/capture_pipeline/bentley.py
from __future__ import annotations

from pathlib import Path
from typing import Callable, Iterable

import httpx
from azure.storage.blob import ContainerClient

API_BASE = "https://api.bentley.com"
ACCEPT = "application/vnd.bentley.itwin-platform.v1+json"


class BentleyApiError(RuntimeError):
    pass


class BentleyClient:
    def __init__(self, access_token: str, *, timeout: float = 60.0) -> None:
        if not access_token.strip():
            raise ValueError("Bentley access token is empty")
        self._client = httpx.Client(
            base_url=API_BASE,
            timeout=timeout,
            headers={
                "Authorization": f"Bearer {access_token.strip()}",
                "Accept": ACCEPT,
                "Content-Type": "application/json",
            },
        )

    def close(self) -> None:
        self._client.close()

    def __enter__(self) -> "BentleyClient":
        return self

    def __exit__(self, exc_type: object, exc: object, tb: object) -> None:
        self.close()

    def _json(self, response: httpx.Response) -> dict[str, object]:
        try:
            response.raise_for_status()
        except httpx.HTTPStatusError as exc:
            detail = response.text.strip()
            raise BentleyApiError(
                f"Bentley API returned {response.status_code}: {detail or response.reason_phrase}"
            ) from exc
        payload = response.json()
        if not isinstance(payload, dict):
            raise BentleyApiError("Bentley API returned an unexpected response")
        return payload

    def create_image_collection(self, *, itwin_id: str, display_name: str) -> dict[str, object]:
        payload = self._json(
            self._client.post(
                "/reality-management/reality-data/",
                json={
                    "iTwinId": itwin_id,
                    "displayName": display_name,
                    "classification": "Undefined",
                    "type": "CCImageCollection",
                    "authoring": True,
                },
            )
        )
        reality_data = payload.get("realityData", payload)
        if not isinstance(reality_data, dict) or not reality_data.get("id"):
            raise BentleyApiError("Create reality data response did not include an id")
        return reality_data

    def get_write_container_url(self, *, reality_data_id: str, itwin_id: str) -> str:
        payload = self._json(
            self._client.get(
                f"/reality-management/reality-data/{reality_data_id}/writeaccess",
                params={"iTwinId": itwin_id},
            )
        )
        container = payload.get("container", payload)
        try:
            return str(container["_links"]["containerUrl"]["href"])  # type: ignore[index]
        except (KeyError, TypeError) as exc:
            raise BentleyApiError("Write-access response did not include a container URL") from exc

    def finalize_reality_data(self, *, reality_data_id: str, itwin_id: str) -> dict[str, object]:
        payload = self._json(
            self._client.patch(
                f"/reality-management/reality-data/{reality_data_id}",
                json={"iTwinId": itwin_id, "authoring": False},
            )
        )
        reality_data = payload.get("realityData", payload)
        if not isinstance(reality_data, dict):
            raise BentleyApiError("Finalize response was not a reality data object")
        return reality_data


def upload_files(
    *,
    container_url: str,
    files: Iterable[Path],
    root: Path,
    prefix: str = "",
    on_uploaded: Callable[[Path, str], None] | None = None,
) -> int:
    resolved_root = root.expanduser().resolve()
    prefix = prefix.strip("/")
    container = ContainerClient.from_container_url(container_url)
    uploaded = 0

    for path in files:
        resolved = path.expanduser().resolve()
        if resolved_root.is_dir():
            relative = resolved.relative_to(resolved_root).as_posix()
        else:
            relative = resolved.name
        blob_name = f"{prefix}/{relative}" if prefix else relative

        with resolved.open("rb") as stream:
            container.upload_blob(name=blob_name, data=stream, overwrite=False)
        uploaded += 1
        if on_uploaded is not None:
            on_uploaded(resolved, blob_name)

    return uploaded
